map/bookmark_helpers.cpp

```diff
@@ -370,33 +371,40 @@ std::string GetKMLorGPXPath(std::string const & filePath)
     if (kmlData == nullptr)
       return {};
 
-    fileSavePath = GenerateValidAndUniqueFilePathForKML(base::FileNameFromFullPath(filePath));
+    auto fileSavePath = GenerateValidAndUniqueFilePathForKML(base::FileNameFromFullPath(filePath));
     if (!SaveKmlFileByExt(*kmlData, fileSavePath))
       return {};
+    return {std::move(fileSavePath)};
   }
   else if (fileExt == kKmzExtension)
   {
-    // Extract KML file from KMZ archive and save to temp KML with unique name.
+    // Extract KML files from KMZ archive and save to temp KMLs with unique name.
+    std::vector<std::string> kmlFilePaths;
     try
     {
       ZipFileReader::FileList files;
       ZipFileReader::FilesList(filePath, files);
-      std::string kmlFileName;
-      std::string ext;
-      for (size_t i = 0; i < files.size(); ++i)
+      std::string indexFileName;
+      std::vector<std::string> kmlFileNames;
+      for (const auto & file : files)
       {
-        ext = GetLowercaseFileExt(files[i].first);
+        auto const & kmlFileName = file.first;
+        std::string ext = GetLowercaseFileExt(kmlFileName);
         if (ext == kKmlExtension)
-        {
-          kmlFileName = files[i].first;
-          break;
-        }
+          kmlFileNames.push_back(kmlFileName);
       }
-      if (kmlFileName.empty())
-        return {};
-
-      fileSavePath = GenerateValidAndUniqueFilePathForKML(kmlFileName);
-      ZipFileReader::UnzipFile(filePath, kmlFileName, fileSavePath);
+      for (const auto & kmlFileName : kmlFileNames)
+      {
+        auto fileSavePath = GenerateValidAndUniqueFilePathForKML(kmlFileName);
+        ZipFileReader::UnzipFile(filePath, kmlFileName, fileSavePath);
+        auto start = kmlFileName.find_last_of('/') + 1;
+        auto const name = kmlFileName.substr(start, kmlFileName.size() - start);
+        // skip index kml if there are more than 1 kml files in kmz
+        if (name == "doc.kml" && kmlFileNames.size() > 1)
+          continue;
+        kmlFilePaths.push_back(std::move(fileSavePath));
+      }
+      return kmlFilePaths;
```

Comment:
I didn't get how it could happen - in case of exception it will be handled in this block
```
    catch (RootException const & e)
    {
      LOG(LWARNING, ("Error unzipping file", filePath, e.Msg()));
      return {};
    }
```
and empty result is returned?

Reply:
If KML has 2+ files, and the first one will be successfully unpacked from zip, but the next one will fail (e.g.  ZipFileReader::UnzipFile throws an exception), then `return {}` won't return the first file that was successfully extracted.

Reply:
ah, got it.
Fixed.